src/xbot/common/command/scripted/ScriptedCommand.java

```diff
@@ -0,0 +1,222 @@
+     * For internal use only!
+     * @param command The command to run
+     */
+    public synchronized void invokeCommand(BaseCommand command) {
+        log.info("Adding command " + command + " to scheduler");
+        
+        CancellableCommandWrapper wrappedCommand = new CancellableCommandWrapper(command);
+        
+        this.scheduler.add(wrappedCommand);
+        if(this.invokedCommands == null) {
+            this.invokedCommands = new HashSet<BaseCommand>();
+        }
+        
+        this.invokedCommands.add(wrappedCommand);
+    }
+    
+    @Override
+    public void initialize() {
+        log.info("Initialized");
+        initializeExecThread();
+    }
+
+    @Override
+    public void execute() {
+        // Start the execution thread if this is the first command run
+        if(execThread != null && execThread.getState() == State.NEW) {
+            log.info("Starting exec thread");
+            this.execThread.start();
+        }
+        
+        // Update any locks waiting for commands to finish
+        if(execThread != null && this.completionLocks != null) {
+            Set<ScriptedCommandCompletionLock> locksToRemove = new HashSet<>();
+            
+            for(ScriptedCommandCompletionLock lock : this.completionLocks) {
+                if(lock.updateLock()) {
+                    log.debug("Found lock to remove: " + lock.toString());
+                    locksToRemove.add(lock);
+                }
+            }
+            
+            if(!locksToRemove.isEmpty()) {
+                log.info("Removing " + locksToRemove.size() + " locks");
+                completionLocks.removeAll(locksToRemove);
+            }
+        }
+    }
+
+    @Override
+    public boolean isFinished() {
+        log.debug("isFinished called; Thread state is " + (execThread == null ? "[uninitialized]" : execThread.getState().toString()));
+        
+        boolean isFinished = execThread != null && execThread.getState() == State.TERMINATED;
+        log.debug("isFinished is " + isFinished);
+        
+        return isFinished;
+    }
+
+    @Override
+    public void end() {
+        log.info("Ending");
+        if(execThread != null) {
+            log.debug("Interrupting exec thread");
+            this.execThread.interrupt();
+        }
+        
+        log.debug("Cancelling invoked commands");
+        if(this.invokedCommands != null) {
+            for(BaseCommand command : invokedCommands) {
+                log.info("Cancelling command " + command);
+
+                command.cancel();
+            }
+            
+            this.invokedCommands = null;
+        }
+        else {
+            log.info("No invoked commands to cancel");
+        }
+    }
+
+    @Override
+    public void interrupted() {
+        log.info("Interrupted");
+        end();
+    }
+    
+    public boolean hasReachedCheckpoint(String checkpointName) {
+        return execThread != null && execThread.hasReachedCheckpoint(checkpointName);
```

Comment:
No; it's a feature -- the idea is that if the script is going to do something in parallel with something else in our system it can signal when certain things happen. Although it _is_ most useful in tests.